### PR TITLE
[AMORO-1983] Upgrading Flink 1.17 based on Mixed-format table

### LINKED ISSUE #1983: [Feature]: Upgrading Flink version based on Mixed-format table
### Description

In terms of [Disscussion-1937](https://github.com/NetEase/amoro/discussions/1937), we conclude that flink modules will extract the public modules to support different versions of the Flink, and will only keep the three latest flink versions in the future, which will greatly reduce maintenance costs and make UTs easier!

The code of the flink public modules will depend on the latest flink version, e.g. 1.17 now, and any conflicts between api's of different Flink versions will be resolved in other versions.

/flink
|---/v1.15
|------/flink #fix the confict with 1.17 API
|------/flink-runtime #shaded 
|---/v1.16
|------/flink #fix the confict with 1.17 API
|------/flink-runtime #shaded 
|---/v1.17
|------/flink 
|------/flink-runtime #shaded 
|---/flink-common #dependence the newest flink version 1.17
|---/flink-common-format #dependence the newest flink version 1.17

### Use case/motivation

_No response_

### Describe the solution

- Support Flink 1.17 based on the Flink common module
- Support Flink 1.16 based on the Flink common module
- Remove Flink 1.12
- Remove Flink 1.14
- Adjust Flink 1.15 based on the Flink common module
### Subtasks

_No response_

- [x] #2059

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by czy006**
## Why are the changes needed?

Close #1983.

## Brief change log

- Copy Flink 1.15 Code to Flink-Common Mode And Flink-Common-format
- Support Version Flink 1.17 and
- Fix Wrong Test Api
- Remove KafkaTestBase （Just One Kafka Test Util）

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible
- [x] Add screenshots for manual tests if appropriate
- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by czy006** [comment]
image

image

These are some errors during unit testing. It can be seen that some packages are missing.

 **org.apache.calcite.rel.metadata.DefaultRelMetadataProvider**

It comes from **flink-table-planner_2.12-1.17** I confirmed that it has been imported but there is still an error. I am looking for the reason.

**Comment by YesOrNo828** [comment]
Added Spotless plugin to Flink module via [PR](https://github.com/NetEase/amoro/pull/2019), @czy006. Please review and add to the flink-common and flink-1.17 modules.

**Comment by czy006** [comment]
> Added Spotless plugin to Flink module via [PR](https://github.com/NetEase/amoro/pull/2019), @czy006. Please review and add to the flink-common and flink-1.17 modules.

Get It

**Comment by czy006** [comment]
image

Use Quick Start Mixed Format Demo with Flink Local Test

File: flink/flink-common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+                    <groupId>org.ow2.asm</groupId>
+                    <artifactId>asm</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <dependency>
+            <groupId>com.netease.amoro</groupId>
+            <artifactId>amoro-hive</artifactId>
+            <version>${project.version}</version>
```
Comment (zhoujinsong) [inline flink/flink-common/pom.xml:58]: The `version` tag here seems to be omited.

File: flink/flink-common/pom.xml [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+
+        <dependency>
+            <groupId>cglib</groupId>
+            <artifactId>cglib</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.9.0</version>
```
Comment (zhoujinsong) [inline flink/flink-common/pom.xml:101]: I suggest using Properties to manage the versions of dependencies.

File: flink/flink-common/pom.xml [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+                    <artifactId>slf4j-api</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <!--   exclude jackson-databind   -->
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.10.2</version>
```
Comment (zhoujinsong) [inline flink/flink-common/pom.xml:203]: The same here.

File: flink/flink-common/pom.xml [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+            <groupId>org.testcontainers</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>${testcontainers.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
```
Comment (zhoujinsong) [inline flink/flink-common/pom.xml:392]: The scope can be inherited from the parent and omitted.

File: flink/pom.xml [hunk 5]
```diff
@@ -41,6 +41,11 @@
         <module>v1.15/flink-format</module>
         <module>v1.15/flink-runtime</module>
         <module>v1.15/arctic-flink-1.15-iceberg-bridge</module>
+        <module>v1.17/flink</module>
+        <module>v1.17/flink-runtime</module>
+        <module>v1.17/amoro-flink-1.17-iceberg-bridge</module>
```
Comment (zhoujinsong) [inline flink/pom.xml:46]: Can we move the classes in `iceberg-bridge` into `flink-common` rather than maintaining a separate copy for each version.

**Review by zhoujinsong — COMMENTED** [review]
@czy006 Thanks a lot for your work!

I left some commets.
BTW, we may need to maintain additional integration test cases for each Flink version. At this stage, we can select appropriate test cases from "common" and migrate them over. What do you think?

File: flink/pom.xml (hunk 5 above)
Comment (czy006) [reply flink/pom.xml]: now move to new model **flink-common-iceberg-bridge**

**Comment by czy006** [comment]
> @czy006 Thanks a lot for your work!
> 
> I left some commets. BTW, we may need to maintain additional integration test cases for each Flink version. At this stage, we can select appropriate test cases from "common" and migrate them over. What do you think?

We should differentiate between these unit tests. After communication, I think some unit tests should be integrated under each Flink version to ensure that the functions of the current version can be covered

**Comment by czy006** [comment]
after discuss ,patch #1994 feature in flink 1.17

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**